store/driver/txn/snapshot.go

```diff
@@ -28,28 +28,66 @@ import (
 
 type tikvSnapshot struct {
 	*txnsnapshot.KVSnapshot
+	// customRetrievers stores all custom retrievers, it is sorted
+	customRetrievers sortedRetrievers
 }
 
 // NewSnapshot creates a kv.Snapshot with txnsnapshot.KVSnapshot.
 func NewSnapshot(snapshot *txnsnapshot.KVSnapshot) kv.Snapshot {
-	return &tikvSnapshot{snapshot}
+	return &tikvSnapshot{snapshot, nil}
 }
 
 // BatchGet gets all the keys' value from kv-server and returns a map contains key/value pairs.
 // The map will not contain nonexistent keys.
 func (s *tikvSnapshot) BatchGet(ctx context.Context, keys []kv.Key) (map[string][]byte, error) {
+	customData := make(map[string][]byte)
```

Comment:
`if s.customRetrievers != nil` ?

Reply:
This code path is performance sensitive, avoid any unnecessary allocation please.